project/src/ExternalInterface.cpp

```diff
@@ -133,15 +135,82 @@ namespace lime {
 	}
 
 
+	std::string wstring_utf8 (const std::wstring& val) {
+
+		std::string out;
+		unsigned int codepoint = 0;
+
+		for (const wchar_t chr : val) {
+
+			if (chr >= 0xd800 && chr <= 0xdbff) {
+
+				codepoint = ((chr - 0xd800) << 10) + 0x10000;
+
+			} else {
+
+				if (chr >= 0xdc00 && chr <= 0xdfff) {
+
+					codepoint |= chr - 0xdc00;
+
+				} else {
+
+					codepoint = chr;
+
+				}
+
+				if (codepoint <= 0x7f) {
+
+					out.append (1, static_cast<char> (codepoint));
+
+				} else if (codepoint <= 0x7ff) {
+
+					out.append (1, static_cast<char> (0xc0 | ((codepoint >> 6) & 0x1f)));
+					out.append (1, static_cast<char> (0x80 | (codepoint & 0x3f)));
+
+				} else if (codepoint <= 0xffff) {
+
+					out.append (1, static_cast<char> (0xe0 | ((codepoint >> 12) & 0x0f)));
+					out.append (1, static_cast<char> (0x80 | ((codepoint >> 6) & 0x3f)));
+					out.append (1, static_cast<char> (0x80 | (codepoint & 0x3f)));
+
+				} else {
+
+					out.append (1, static_cast<char> (0xf0 | ((codepoint >> 18) & 0x07)));
+					out.append (1, static_cast<char> (0x80 | ((codepoint >> 12) & 0x3f)));
+					out.append (1, static_cast<char> (0x80 | ((codepoint >> 6) & 0x3f)));
+					out.append (1, static_cast<char> (0x80 | (codepoint & 0x3f)));
+
+				}
+
+				codepoint = 0;
+
+			}
+
+		}
+
+		return out;
+
+	}
+
+
+	vbyte* hl_wstring_to_utf8_bytes (const std::wstring& val) {
+
+		const std::string utf8 (wstring_utf8 (val));
+		vbyte* const bytes = hl_alloc_bytes (utf8.size () + 1);
+		std::memcpy(bytes, utf8.c_str (), utf8.size () + 1);
+		return bytes;
+
+	}
+
+
 	std::wstring* hxstring_to_wstring (HxString val) {
 
 		if (val.c_str ()) {
 
-			std::string _val = std::string (val.c_str ());
 			#ifdef HX_WINDOWS
-			std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
-			return new std::wstring (converter.from_bytes (_val));
+			return new std::wstring (hxs_wchar (val, nullptr));
 			#else
+			const std::string _val (hxs_utf8 (val, nullptr));
 			return new std::wstring (_val.begin (), _val.end ());
```

Comment:
Is there a reason to have two separate conversions here for Windows and other platforms? Hxcpp's strings are the same regardless of the platform (same for HL). Or is it creating different wstrings here?

Reply:
The encoding of `std::wstring` is platform-dependent (or even compiler-dependent) but for GCC and MSVC it's UTF-16 on Windows and UTF-32/UCS-4 on Linux and Mac. However that dosen't seem to be really important here because this function `hxstring_to_wstring()` is used only for file dialogs and it's results only as mere transport containers for data to be passed to functions defined in `FileDialog.cpp` which use two sets od APIs. One (Windows-only) that uses UTF-16 and the other one (non-Windows platforms) that uses UTF-8.

Reply:
The situation is a bit messy, but that's for another PR.